File: src/run-vector.ts
import fs from 'node:fs';

import { initState } from './state.js';
import { nextStep } from './next-step.js';

type VectorCycle = { R: number; H: number; O: number };

type VectorFile = {
  name?: string;
  description?: string;
  cardPath: string;
  initialState?: any;
  cycles: VectorCycle[];
};

export function runVector(vectorPath: string) {
  const v: VectorFile = JSON.parse(fs.readFileSync(vectorPath, 'utf8'));
  const card = JSON.parse(fs.readFileSync(v.cardPath, 'utf8'));

  let state = v.initialState ?? initState();

  const out: any[] = [];
  for (let i = 0; i < v.cycles.length; i++) {
    const score = v.cycles[i];
    const res = nextStep(card, state, score as any);
    out.push({
      i: i + 1,
      score,
      action: res.action,
      escalationLevel: res.why.escalationLevel,
      rhoSum: res.why.rhoSum,
      choreography: (res.state as any).choreography
        ? {
            stepIndex: (res.state as any).choreography.stepIndex,
            cyclesInStep: (res.state as any).choreography.cyclesInStep,
            profile: (res.state as any).choreography.profile
          }
        : undefined
    });
    state = res.state as any;
  }

  return {
    name: v.name,
    description: v.description,
    vectorPath,
    cardPath: v.cardPath,
    out
  };
}

File: python/coordcard/run_vector.py
import json
from pathlib import Path
from typing import Any, Dict

from .state import init_state
from .next_step import next_step


def run_vector(vector_path: str) -> Dict[str, Any]:
    v = json.loads(Path(vector_path).read_text(encoding="utf-8"))
    card_path = v["cardPath"]
    card = json.loads(Path(card_path).read_text(encoding="utf-8"))

    state = v.get("initialState") or init_state()

    out = []
    for i, cyc in enumerate(v.get("cycles", [])):
        res = next_step(card, state, cyc)
        why = res.get("why", {}) or {}
        out.append(
            {
                "i": i + 1,
                "score": cyc,
                "action": res.get("action"),
                "templateText": res.get("templateText"),
                "why": {
                    "triggerFired": why.get("triggerFired"),
                    "ruleSource": why.get("ruleSource"),
                    "ruleText": why.get("ruleText"),
                    "summary": why.get("summary"),
                    "rhoSum": why.get("rhoSum"),
                    "escalationLevel": why.get("escalationLevel"),
                },
                "escalationLevel": why.get("escalationLevel"),
                "rhoSum": why.get("rhoSum"),
                "choreography": res.get("state", {}).get("choreography"),
            }
        )
        state = res.get("state")

    return {
        "name": v.get("name"),
        "description": v.get("description"),
        "vectorPath": vector_path,
        "cardPath": card_path,
        "out": out,
    }
